fix: add deprecated overload for Observable based access

iwt-dahlborn opened the pull request:
Currently if you call the perception registrar with a subject that is both `Perceptible` and `Observable` it will call the `Observable` variants since the `Perceptible` implementations are disfavored.

The exception is if your subject still calls the deprecated variant of the access function, in this case the `Perceptible` variant will be called possibly leading to an erroneous perception runtime warning.

Adding a `Observable` based implementation of the deprecated function reroutes the call to the non deprecated `Observable` implementation properly.

I am not sure whether you want to integrate this since it practically adds a deprecated function and it can also be fixed on the user side by moving to the non-deprecated function. But the perception warning lead to some confusion in our project.

Commit:
add deprecated overload for Observable based access

mbrandonw commented:
Hi @iwt-dahlborn, currently it is expected for you to get a perception warning if you are deploying to iOS 16 even if you are currently running on an iOS 17+ device. And the way we have things set up now is what allows that.

Can you confirm that you were definitely receiving a perception warning for a project that was deploying to iOS 17+?

iwt-dahlborn commented:
Yes, the minimum deployment version of the app, and package the call were contained in, is iOS 17.0. 

Aditionally I was able to avoid the perception warning by switching to the non deprecated access call.

mbrandonw commented:
I just may be misunderstanding the root issue, but can you describe why you were calling a deprecated method? Did something about the library force you into that situation?

If this PR is only fixing something that only happens when invoking deprecated APIs, then I don't think we will be merging it. Deprecated APIs should be avoided and not sure we should introduce new surface area to help out those still using deprecated APIs.

iwt-dahlborn commented:
Basically I updated the tca version and migrated to observable but had not gotten around to update the deprecated call in this case.

Before this function was deprecated it had 2 overloads:
```swift
// Perceptible version
public func access<Subject: Perceptible, Member>(
      _ subject: Subject,
      keyPath: KeyPath<Subject, Member>,
      fileID: StaticString = #fileID,
      filePath: StaticString = #filePath,
      line: UInt = #line,
      column: UInt = #column
) {

// Observable version
public func access<Subject: Observable, Member>(
      _ subject: Subject,
      keyPath: KeyPath<Subject, Member>,
      fileID: StaticString = #fileID,
      filePath: StaticString = #filePath,
      line: UInt = #line,
      column: UInt = #column
) {
```

But currently there exists only a deprecation for the `Perceptible` version while the `Observable` version was practically completely removed. But since its an overload any places in the code that still use the deprecated version of the function will call the `Perceptible` version of the implementation if the subject is both Perceptible and Observable (for example in a package that needs to support iOS 16 but also iOS 17+ apps. Even though before the deprecation the Observable implementation was called.

The change in this pr practically ensures that the code works as expected for people who update the sdk and have not gotten around to update all the deprecation (which I know one should do).

And I understand that you might not want to merge it since the fix for the user is to just update the code to the non-deprecated version of the function. Normally re-adding a function that was already removed as a deprecated would not make sense but since its an overloaded function that can under certain circumstances just "silently" (except for the deprecation warning) switches to the other overload it could help people who are updating to a new version of the package.

So this would practically only be to ease the transition. 

Which also of course would only help if they update to this certain patch.

If you think this is not a good idea we can discard this pr. Its not "needed" on my side since I simply removed the deprecated function calls to solve the issue.

Edit: Small correction since its soft deprecated a deprecation warning for this call is not immediately visible in xcode causing extra confusion as to why it causes weird behavior.

iwt-dahlborn commented:
I created a small example project to visualise the "issue" when you open it you wont see any warning or errors in the code, but if you run it you get the perception warning asking you to add `WithPerceptionTracking` if you do that you will get a warning that it isn't necessary for iOS 17+.

So since the function is only soft deprecated its not immediately obvious what you need to do to fix the perception warning, the only way to really find out is to attach the debugger and step through to hopefully see that the function is deprecated and understand that it links to the perceptible implementation instead of the observable one now.

The additional deprecated function in this pr practically just allows the function to stay soft deprecated but still call the observable variant of the implementation.

[Deprecation.zip](https://github.com/user-attachments/files/23602419/Deprecation.zip)

stephencelis commented:
@iwt-dahlborn I believe things are behaving the way we expect. The `access(_:keyPath:fileID:filePath:line:column:)` endpoint you're hitting on `PerceptionRegistrar` will and should always go through the perception checking machinery. If you truly want to disable perception checking on certain runtimes, you must do an availability check and call the `access(_:keyPath:)` method on that branch:

```swift
    nonisolated func access<Member>(
        keyPath: KeyPath<Self, Member>,
        fileID: StaticString = #fileID,
        filePath: StaticString = #filePath,
        line: UInt = #line,
        column: UInt = #column
    ) {
        if #available(iOS 17, macOS 14, tvOS 17, watchOS 10, *) {
            perceptionRegistrar.access(
                self,
                keyPath: keyPath
            )
        } else {
            perceptionRegistrar.access(
                self,
                keyPath: keyPath,
                fileID: fileID,
                filePath: filePath,
                line: line,
                column: column
            )
        }
    }
```

Does this help?

iwt-dahlborn commented:
> I believe things are behaving the way we expect. The `access(_:keyPath:fileID:filePath:line:column:)` endpoint you're hitting on `PerceptionRegistrar` will and should always go through the perception checking machinery.

Ok, maybe I still misunderstand but in `1.6.0` of the perception library the [access(_:keyPath:fileID:filePath:line:column:)](https://github.com/pointfreeco/swift-perception/blob/d924c62a70fca5f43872f286dbd7cef0957f1c01/Sources/PerceptionCore/PerceptionRegistrar.swift#L55) did not call the perception check because it had a dedicated overload for iOS 17+ and observable. While in 2.0 this overload was removed and only the perceptible version was deprecated. 

What you are saying is that this was an issue in 1.6 that was fixed by removing this function?

Regarding the availability check the true fix would just be to call only access without file paths right?
```swift
nonisolated func access<Member>(
        keyPath: KeyPath<Self, Member>,
        fileID: StaticString = #fileID,
        filePath: StaticString = #filePath,
        line: UInt = #line,
        column: UInt = #column
    ) {
        perceptionRegistrar.access(
            self,
            keyPath: keyPath
        )
    }
```

Since it will call the correct overload depending on the iOS version and the implemented protocols.

And the deprecated function discards the file parameters anyway.

This pr was only intended to fix a fairly niche issue that would crop up when updating from 1.6 to 2.x anyway and if it its actually an intended change I understand this not being merged.

Do you think it would make sense to at least change the deprecation notice from:
```swift
@available(iOS, deprecated: 9999, renamed: "access(_:keyPath:)")
```
to
```swift
@available(iOS, deprecated: 17.0, renamed: "access(_:keyPath:)")
```
?

mbrandonw commented:
Hi @iwt-dahlborn, this library has a very delicate web of overloads so it's hard to remember the exact incantation needed to provide the behavior we want. From my perspective, the behavior right now _is_ what we want, and so perhaps that means 1.6 had a hole that allowed for incorrect behavior. Or perhaps we are misremembering how the overloads were balanced in 1.6 and did not faithfully recreate it for 2.0.

But, the behavior we want is that if you are deploying to iOS <17, then we always want to show perception warnings even if you are running on an iOS 17+ device. This is because while running on iOS 17+ your app will work correctly and so you may not know that secretly iOS <17 is broken.

Looking at your sample project I'm not sure I understand why you are using `PerceptionRegistrar` in the first place. The deployment target is iOS 18.6+, and so you can just use `ObservationRegistrar`. And previously you stated that your app + package is iOS 17+. The only time you need `PerceptionRegistrar` is if you are deploying on older runtimes.

Is there a reason you can't update all Perception APIs to Observation? Or are you saying that you are currently in the process of doing that, and in the meantime you rather not have perception runtime warnings?

iwt-dahlborn commented:
@mbrandonw 
> Or are you saying that you are currently in the process of doing that, and in the meantime you rather not have perception runtime warnings?

Yes, practically when we updated to the new perception version and updated the minimum deployment version to iOS 17.0, we triggered this issue and it wasn't immediately obvious why it was happening since the method was only soft deprecated and at first glance all the functions redirect to observation registrar.

This change practically just aimed to make the transition easier by keeping the overload behaviour of 1.6 and matching the other calls in perception registrar. I already solved this issue in my project by not using the deprecated functions.

As I mentioned in the beginning if you think it is not worth doing or doesn't align with the vision this pr can just be closed, I just thought it would be an easy change to smooth out the transition for other people updating their dependencies or minimum deployment versions.

But yeah this discussion might have already been longer than it was worth, sorry for that.
